### PR TITLE
DigitalOcean native support #1594

### PR DESCRIPTION
**Pull request by zbendhiba**
[ ] An issue should be filed for the change unless this is a trivial change (fixing a typo or similar). One issue should ideally be fixed by not more than one commit and the other way round, each commit should fix just one issue, without pulling in other changes.
[ ] Each commit in the pull request should have a meaningful and properly spelled subject line and body. Copying the title of the associated issue is typically enough. Please include the issue number in the commit message prefixed by #.
[ ] The pull request description should explain what the pull request does, how, and why. If the info is available in the associated issue or some other external document, a link is enough.
[ ] Phrases like Fix #<issueNumber> or Fixes #<issueNumber> will auto-close the named issue upon merging the pull request. Using them is typically a good idea.
[ ] Please run mvn process-resources -Pformat (and amend the changes if necessary) before sending the pull request.
[ ] Contributor guide is your good friend: https://camel.apache.org/camel-quarkus/latest/contributor-guide.html

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

**Comment by zbendhiba** [comment]
I've created an issue to add more tests https://github.com/apache/camel-quarkus/issues/2388

File: extensions/digitalocean/deployment/src/main/java/org/apache/camel/quarkus/component/digitalocean/deployment/DigitaloceanProcessor.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        IndexView index = combinedIndex.getIndex();
+        // register Digitalocean model classes for JSON serialize/deserialize
+        String[] dtos = index.getKnownClasses().stream().map(ci -> ci.name().toString())
+                .filter(n -> n.startsWith("com.myjeeva.digitalocean.pojo")
+                        || n.startsWith("com.myjeeva.digitalocean.common"))
+                .sorted()
+                .toArray(String[]::new);
+
+        return new ReflectiveClassBuildItem(true, true, dtos);
+    }
```
Comment (ppalaga) [inline extensions/digitalocean/deployment/src/main/java/org/apache/camel/quarkus/component/digitalocean/deployment/DigitaloceanProcessor.java:53]: Jandex!

File: integration-tests/digitalocean/src/test/java/org/apache/camel/quarkus/component/digitalocean/it/DigitaloceanTest.java [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+
+@QuarkusTest
+@QuarkusTestResource(DigitaloceanTestResource.class)
+class DigitaloceanTest {
+
+    @MockServer
+    WireMockServer server;
+
+    @Test
+    @DisabledIfEnvironmentVariable(named = "DIGITALOCEAN_AUTH_TOKEN", matches = ".+")
```
Comment (ppalaga) [inline integration-tests/digitalocean/src/test/java/org/apache/camel/quarkus/component/digitalocean/it/DigitaloceanTest.java:51]: 1. Disabled if not-empty sounds strange - should that really be that way?
2. You say in the readme "By default, the Digitalocean integration tests use WireMock to stub the API interactions." which sounds like `DIGITALOCEAN_AUTH_TOKEN` is not needed if I want to run against wiremock. But with this `@DisabledIfEnvironmentVariable`, the test does not run at all against wiremock. Or I am misunderstanding something?

**Review by ppalaga — COMMENTED** [review]
A question inline. Otherwise looks good.

File: integration-tests/digitalocean/src/test/java/org/apache/camel/quarkus/component/digitalocean/it/DigitaloceanTest.java (hunk 2 above)
Comment (zbendhiba) [reply integration-tests/digitalocean/src/test/java/org/apache/camel/quarkus/component/digitalocean/it/DigitaloceanTest.java:51]: Maybe the main test, I just have to get the same timeout
But for the 2 other tests : Kernels and Neighbors, I couldn't get the responses with my basic test from Digitalocean. I did a mock using their examples

Comment (zbendhiba) [reply integration-tests/digitalocean/src/test/java/org/apache/camel/quarkus/component/digitalocean/it/DigitaloceanTest.java:51]: FYI, I'm updating this

**Comment by ppalaga** [comment]
Master is on 1.9.0-SNAPSHOT now. Could you plz rebase and change the since native version to 1.9.0?

File: extensions/digitalocean/deployment/src/main/java/org/apache/camel/quarkus/component/digitalocean/deployment/DigitaloceanProcessor.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    @BuildStep
+    ReflectiveClassBuildItem registerForReflection(CombinedIndexBuildItem combinedIndex) {
+        IndexView index = combinedIndex.getIndex();
+        // register Digitalocean model classes for JSON serialize/deserialize
+        String[] dtos = index.getKnownClasses().stream().map(ci -> ci.name().toString())
+                .filter(n -> n.startsWith("com.myjeeva.digitalocean.pojo")
+                        || n.startsWith("com.myjeeva.digitalocean.common"))
+                .sorted()
```
Comment (aldettinger) [inline extensions/digitalocean/deployment/src/main/java/org/apache/camel/quarkus/component/digitalocean/deployment/DigitaloceanProcessor.java:49]: Maybe we don't need sort them ?

File: integration-tests-support/wiremock/src/main/java/org/apache/camel/quarkus/test/wiremock/WireMockTestResourceLifecycleManager.java [hunk 4]
```diff
@@ -217,6 +219,9 @@ private WireMockServer createServer() {
             // Read mapping resources from the classpath in playback mode
             configuration.fileSource(new CamelQuarkusFileSource());
         }
+
```
Comment (aldettinger) [inline integration-tests-support/wiremock/src/main/java/org/apache/camel/quarkus/test/wiremock/WireMockTestResourceLifecycleManager.java:222]: I read it like HTTPs is enabled for all tests. It reminds me that we added the possibility to customize the configuration in specific cases with `customizeWiremockConfiguration()`.

Comment (zbendhiba) [reply integration-tests-support/wiremock/src/main/java/org/apache/camel/quarkus/test/wiremock/WireMockTestResourceLifecycleManager.java:222]: @aldettinger thanks for the tips. I'll use that!!

File: integration-tests/digitalocean/src/main/java/org/apache/camel/quarkus/component/digitalocean/it/DigitaloceanRoute.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+     * We need to implement some conditional configuration of the {@link DigitalOceanClient} thus we create it
+     * programmatically
+     *
+     * @return a configured {@link DigitalOceanClient}
+     */
+    @Produces
+    @ApplicationScoped
+    @Unremovable
+    @Named("digitalOceanClient")
+    DigitalOceanClient intiDigitalOceanClient(MockApiService mockApiService) {
```
Comment (aldettinger) [inline integration-tests/digitalocean/src/main/java/org/apache/camel/quarkus/component/digitalocean/it/DigitaloceanRoute.java:43]: small typo here ? init

File: integration-tests/digitalocean/src/main/java/org/apache/camel/quarkus/component/digitalocean/it/DigitaloceanRoute.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            digitalOceanClient.setApiVersion("V2");
+            return digitalOceanClient;
+        }
+        return new DigitalOceanClient(oAuthToken);
+    }
+
+    @Override
+    public void configure() throws Exception {
+        from("direct:droplet")
+                .to(String.format("digitalocean:droplets?oAuthToken=%s&digitalOceanClient=#digitalOceanClient", oAuthToken));
```
Comment (aldettinger) [inline integration-tests/digitalocean/src/main/java/org/apache/camel/quarkus/component/digitalocean/it/DigitaloceanRoute.java:57]: For the purpose of sharing tips, toF might be a good shortcut in such cases.

**Comment by aldettinger** [comment]
This is quite a PR. It's on the right track :+1:

**Comment by zbendhiba** [comment]
@ppalaga @aldettinger can we merge this one? Are you okay with the changes? The failing test is related to camel-quarkus-examples quarkus.log file

**Pull request merged**